integrations/ad_reader.py

```diff
@@ -166,14 +151,57 @@ def read_user(self, user=None, cpr=None):
             brugertype = current_user.get('xBrugertype')
             if brugertype and brugertype.find('Medarbejder') == -1:
                 continue
+            if not current_user:
+                current_user = {}
             self.results[current_user['xAttrCPR']] = current_user
             self.results[current_user['SamAccountName']] = current_user
+        return current_user
+
+    def read_user(self, user=None, cpr=None):
+        """
+        Read all properties of an AD user. The user can be retrived either by cpr
+        or by AD user name.
+        :param user: The AD username to retrive.
+        :param cpr: cpr number of the user to retrive.
+        :return: All properties listed in AD for the user.
+        """
+        if (not cpr) and (not user):
+            return
+
+        if user:
+            dict_key = user
+            if user in self.results:
+                return self.results[user]
+
+        if cpr:
+            dict_key = cpr
+            if cpr in self.results:
+                return self.results[cpr]
+
+        m = hashlib.sha256()
+        m.update(dict_key.encode())
+        path_url = 'ad_' + m.hexdigest()
+
+        try:
+            with open(path_url + '.p', 'rb') as f:
+                logger.debug('{} was found in cache'.format(dict_key))
+                response = pickle.load(f)
+                if not response:
+                    response = {}
+                self.results[dict_key] = response
+
+        except FileNotFoundError:
+            response = self.uncached_read_user(user=user, cpr=cpr)
+            with open(path_url + '.p', 'wb') as f:
+                pickle.dump(response, f, pickle.HIGHEST_PROTOCOL)
+
+        logging.debug('Returned info for {}'.format(dict_key))
```

Comment:
Endelig noget, jeg har lidt forstand på.:
Jeg synes du skal bruge den fine 'logger' du har lavet i stedet for modulet 'logging', bare for at få logger-navnet med ud på linjen. Ellers ved jeg såmænd ikke om der er den strore forskel.